zerver/lib/message.py

```diff
@@ -270,24 +270,27 @@ def sew_submessages_and_reactions_to_msgs(messages: List[Dict[str, Any]]) -> Lis
         return sew_messages_and_reactions(messages, reactions)
 
     @staticmethod
-    def to_dict_uncached(messages: List[Message]) -> Dict[int, bytes]:
-        messages_dict = MessageDict.to_dict_uncached_helper(messages)
+    def to_dict_uncached(messages: List[Message], realm_id: Optional[int]=None) -> Dict[int, bytes]:
+        messages_dict = MessageDict.to_dict_uncached_helper(messages, realm_id)
         encoded_messages = {msg['id']: stringify_message_dict(msg) for msg in messages_dict}
         return encoded_messages
 
     @staticmethod
-    def to_dict_uncached_helper(messages: List[Message]) -> List[Dict[str, Any]]:
+    def to_dict_uncached_helper(messages: List[Message],
+                                realm_id: Optional[int]=None) -> List[Dict[str, Any]]:
         # Near duplicate of the build_message_dict + get_raw_db_rows
         # code path that accepts already fetched Message objects
         # rather than message IDs.
 
-        # TODO: We could potentially avoid this database query in
-        # common cases by optionally passing through the
-        # stream_realm_id through the code path from do_send_messages
-        # (where we've already fetched the data).  It would involve
-        # somewhat messy plumbing, but would probably be worth it.
         def get_rendering_realm_id(message: Message) -> int:
-            if message.recipient.type == Recipient.STREAM:
+            if realm_id is not None:
+                return realm_id
+            if (message.recipient.type == Recipient.STREAM) and (message.sender.is_bot is True):
```

Comment:
I'm not sure we want to bake in here the assumption that all cross-realm users are bots.

There's a future world that we will probably end up building that involves there being certain cross-realm streams.